src/System.Buffers.Primitives/System/Buffers/ReaderWriter/BufferReader.cs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Sequences;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public class BufferReader
+    {
+        public static BufferReader<TSequence> Create<TSequence>(TSequence buffer) where TSequence : ISequence<ReadOnlyMemory<byte>>
+        {
+            return new BufferReader<TSequence>(buffer);
+        }
+    }
+
+    public ref partial struct BufferReader<TSequence> where TSequence : ISequence<ReadOnlyMemory<byte>>
+    {
+        private TSequence _sequence;
+        private Position _currentSegmentPosition;
+        private Position _nextSegmentPosition;
+
+        private ReadOnlySpan<byte> _currentSegment;
+        private int _currentSegmentConsumedBytes;
+        
+        // TODO: this should be long, but it will impact the HTTP parser. We should meassure perf impact in isolation
+        private int _totalConsumedBytes;
+        private bool _hasCompleted;
+
+        public BufferReader(TSequence buffer)
+        {
+            _hasCompleted = false;
+            _currentSegmentConsumedBytes = 0;
+            _totalConsumedBytes = 0;
+            _sequence = buffer;
+            _currentSegmentPosition = _sequence.Start;
+            _nextSegmentPosition = _currentSegmentPosition;
+            _currentSegment = default;
+            MoveNext();
+        }
+
+        public bool End => _hasCompleted;
+
+        public int CurrentSegmentIndex => _currentSegmentConsumedBytes;
+
+        public Position CurrentPosition => _currentSegmentPosition + _currentSegmentConsumedBytes;
+
+        public ReadOnlySpan<byte> CurrentSegment => _currentSegment;
```

Comment:
`CurrentSpan`?

Reply:
Or just `Span`

Reply:
I think logically the type is a sequence of segments (that's what we call the concept in Position). Span is a type that we happen to use in this API to represent a segment. I would rather keep property names that describe the concept, not the representation.

Reply:
We have memory in `IMemoryList` not `Segment`, `Segment` is just another term wit didn't use before.